WIP: Windows cross compilation via LLVM

Ericson2314 opened the pull request:
###### Motivation for this change

LLVM can do a much better job than GCC via MinGW. In particular, it can do abi-compatable C++ with MSVC, and has more features implemented such proper C++ threads.

I want this because I want Nix to run on Windows, and am continuing @volth's port to make it do so (https://github.com/NixOS/nix/pull/3185). However, I think it's important that we can purely build Nix for all supported platforms, and not ask developers to set up arcane stuff to test their PRs. Therefore I am making this as a self-imposed prerequisite. I need this to do so because Nix uses C++ threads.

Some of this stuff is rather rougher / and more patch heavy than I like my portability initiatives to be, and so I'll hold off trying to port many more packages until the far far future when the requisite tech debt is cleaned up and the portability is maintainable. But I can't do any less than this if I am to build Nix.

The base branch is very intentional, because I will want to merge to 19.09 in addition to master so Nix can use it from there in its CI.

This supports MinGW and Visual Studio headers and libs:
 - MinGW: `{ config = "x86_64-pc-mingw32"; useLLVM = true; }`
 - MSVC: `{ config = "x86_64-pc-windows-msvc"; useLLVM = true; }`

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @angerman

Ericson2314 committed:
Windows cross with LLVM: Init

[1] Proved incredably useful. Thanks @est31!

[1]: https://github.com/est31/msvc-wine-rust/blob/master/dl.sha256

Ericson2314 committed:
requireFile: Support pname and version, too

Ericson2314 committed:
bintools-wrapper: Support ld64.lld

This is useful if we ever use lld for Darwin instead of cctools.

Ericson2314 committed:
zlib: Build in various ways for Windows

Ericson2314 committed:
Visual Studio and Windows SDKs: Hack case normalize

Ericson2314 committed:
WIP: aws-*: Allow windows builds

Ericson2314 committed:
WIP: openssl: clang for windows support

volth commented:
`pkgs/os-specific/windows` is @aszlig's realm, not mine

I just wanted to say, although LLVM looks promising as the replacement for both GCC and MSVC, no one the latter couple cannot be ruled out.
When you build complex apps (I managed to build native Windows `chromium` so all its buildtime and runtime dependencies) you will need all three.
The Windows-born apps usually need MSVC, ported from Unix - GCC and those which take care of portability (like Chromium itself) are compiled with LLVM (although can be build with MSVC too).
So, ironically, LLVM is the only which could be dropped :)

volth commented:
Also, as far as I remember, LLVM does not include Windows SDK and MinGW has reduced version, so Windows SDK's from microsoft.com is almost inevitable dependency for many Windows apps

Ericson2314 commented:
@volth that's fair, and we could run MSVC in wine too, but I think Nix and all its dependencies, like chromium, should be robust enough to build with LLVM.

matthewbauer commented on lib/systems/examples.nix:
```diff
@@ -220,6 +220,16 @@ rec {
     platform = {};
   };
 
+  windowsLlvm32 = {
```
We should add this to cross-trunk

matthewbauer commented on pkgs/build-support/bintools-wrapper/default.nix:
```diff
@@ -156,9 +156,11 @@ stdenv.mkDerivation {
+    + ''
 
-      for variant in ld.gold ld.bfd ld.lld; do
+      for variant in ld.gold ld.bfd ld.lld${optionalString (targetPlatform.useLLVM or false) " ld64.ldd"}; do
```
I think this wasn’t actually necessary… let’s leave it out for now (perhaps later we can get macOS llvm cross working)

Ericson2314 replied:
Yeah it's not necessary; I was just too busy fixing other things to move it out yet.

matthewbauer commented on pkgs/build-support/cc-wrapper/default.nix:
```diff
@@ -339,6 +339,8 @@ stdenv.mkDerivation {
       hardening_unsupported_flags+=" pic"
     '' + optionalString targetPlatform.isMinGW ''
       hardening_unsupported_flags+=" stackprotector"
+    '' + optionalString targetPlatform.isWindows ''
```
We have so many of these we should provably move these to some attrset (not necessary now)

matthewbauer commented on pkgs/build-support/fetchurl/default.nix:
```diff
@@ -1,4 +1,4 @@
-{ lib, stdenvNoCC, curl }: # Note that `curl' may be `null', in case of the native stdenvNoCC.
+{ lib, buildPackages ? { inherit stdenvNoCC; }, stdenvNoCC, curl }: # Note that `curl' may be `null', in case of the native stdenvNoCC.
```
This seems awkward?

Ericson2314 replied:
Indeed it is!

matthewbauer commented on pkgs/development/libraries/aws-c-common/default.nix:
```diff
@@ -11,6 +11,12 @@ stdenv.mkDerivation rec {
   };
 
+  # It disables it with MSVC, but CMake doesn't realize that clang for windows,
+  # even with gcc-style flags, has the same semantic restrictions.
```
Is aws manually adding fpic? Perhaps they have a flag to disable it.

Ericson2314 replied:
They are manually adding it for non-`MSVC`. CMake should have a better notion of `MSVC` / destinguish between MSVC the abi (which we are) and MSVC which we aren't since the wrappers just know gcc-style flags.

matthewbauer commented on pkgs/os-specific/windows/sdks/default.nix:
```diff
@@ -0,0 +1,264 @@
+{ stdenvNoCC, fetchurl, lib, config, msitools, unzip, static ? false }:
+
+#if !(config.microsoftVisualStudioLicenseAccepted or false)
+if false
```
?

Ericson2314 replied:
Temporary while I get things to work. We should add another line of defense beyond things being broken and not part of the default jobs so @GrahamcOfBorg doesn't do anything with those builds.

matthewbauer commented:
Great work!!!

Ericson2314 commented:
I drafted https://reviews.llvm.org/D69760 which hopefully will fix up some things. Next step might be making a `windres`-style CLI to `llvm-rc`, but hopefully that can be put off for a while.

matthewbauer commented on pkgs/stdenv/cross/default.nix:
```diff
@@ -66,7 +66,7 @@ in lib.init bootStages ++ [
              [ buildPackages.patchelf ]
         ++ lib.optional
-             (let f = p: !p.isx86 || p.libc == "musl" || p.libc == "wasilibc" || p.isiOS; in f hostPlatform && !(f buildPlatform))
+             (let f = p: !p.isx86 || p.libc == "musl" || p.libc == "wasilibc" || p.isiOS || p.isWindows; in f hostPlatform && !(f buildPlatform))
```
I was wondering if we should just set this to `stdenv.hostPlatform != stdenv.buildPlatform`. It very rarely hurts anything (there could be some weird custom gnuconfig changes that break) to set it, we mostly just need it for weird platforms though.

Ericson2314 replied:
Oh! I think I remembered some of the logic was to avoid adding it twice, if native already has it.

We should just do a mass rebuild so everywhere sets it, always.

matthewbauer replied:
I don’t think it’s necessary for native. Besides costing a mass rebuild, it also could break some custom scripts.

Ericson2314 replied:
@matthewbauer We should make a `dontAutoreconf` for that.

matthewbauer replied:
I'd prefer to not have an option for this. Status quo is probably okay. You can always add the hook yourself.

Ericson2314 replied:
Sorry, I meant `dontUpdateAutotoolsGnuConfigScripts`. I we'll only need to add it in a few places.

Ericson2314 committed:
lld 8: Support --allow-multiple-definitions in MinGW frontend

This is needed for building C++ with clang with ld.lld rather than
lld-link.

aszlig commented:
@volth: It isn't my realm anymore, since it has been ages since I had anything to do with Windows (luckily :-D), so I can't really say anything on this PR.

Ericson2314 committed:
clang 8: Allow both ld.lld and lld-link to be used for windows builds

Ericson2314 committed:
lld 8: Backport commit to fix library lookup

We should look for MSVC- and GNU-style names/exensions.

Ericson2314 commented:
Finally zlib builds! But now CMake things don't configure, however (with the switch from `lld-link` to `ld.link`). `INFO:dialect_default[]` it gets running strings on a built binary in both cases. In the `lld-link` case it doesn't care, presumably because there is a "simulating MSVC" thing and the rules are all different. In the `lld.link` / `clang::toolchains::CrossWindows` case it does care, presumably cause it thinks it's more GNU like / MinGW or whatever.

My response is to try going full MinGW, like https://github.com/mstorsjo/llvm-mingw/issues/57, for now, so I am doing that's been done before and which is closer to what we've got.

Ericson2314 committed:
mingw-w64: Patch with master commit so Clang works

Thanks to @mstorsjo in https://github.com/mstorsjo/llvm-mingw/issues/3
for having the answer.

Ericson2314 committed:
Visual Studio and Windows SDKs: Init at 15

Ericson2314 committed:
llvm 8 binutils: Add windres wrapper from llvm-mingw

Ericson2314 committed:
mingw-w64: Allow parallel buildls

Ericson2314 committed:
WIP: libc++ 8: Support Windows

Mic92 commented on pkgs/development/compilers/llvm/8/clang/msvc-lld.patch:
```diff
@@ -0,0 +1,89 @@
+    which might be the first fully automated way to cross compile to
+    Windows.  I haven't taught Nixpkgs to wrap clang-cl like it wraps clang,
+    so even for some packages that have separate MSVC and MinGW builds
+    systems (e.g. zlib, opensll), I seem to be having better luck with the
```
```suggestion
    systems (e.g. zlib, openssl), I seem to be having better luck with the
```

cidkidnix commented:
Heads up here, I'm intending to pick this PR up and get it in a working state against master (or staging-next/staging). 
Currently tinkering with it atm, Would love to have llvm-mingw cross-compiling available.

s1341 commented:
@cidkidnix did you pick this up?

cidkidnix commented:
@s1341 Yeah I did, been super busy lately, but I had something roughly working in this branch: https://github.com/cidkidnix/nixpkgs/commits/dylang/llvm-mingw. Some logic is weird though. I plan to continue on this sometime soon (hopefully).

s1341 commented:
@cidkidnix care for some collaboration? I really need this.

Ericson2314 commented:
@s1341 I think neither of us would mind if you just started picking this up. We can worry about baton passes later.

Also btw check out https://github.com/NixOS/nixpkgs/pull/228374#pullrequestreview-1404603260. It is convenient that UEFI work is also going on concurrently!

Eveeifyeve commented:
Can pr can be closed??? Because the sdk stuff has been revised/done in other prs.
